heron/packing/src/java/com/twitter/heron/packing/builder/HomogeneityScorer.java

```diff
@@ -0,0 +1,51 @@
+// Copyright 2016 Twitter. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package com.twitter.heron.packing.builder;
+
+import com.twitter.heron.spi.packing.PackingPlan;
+
+/**
+ * Scores based on how homogeneous a container is relative to a given component, from highest to
+ * lowest. So a container with all the same components (e.g., AAAA) will get the highest score for
+ * A, while a container with a mix of many components (e.g. ABCD) will get a low score.
+ */
+public class HomogeneityScorer implements Scorer<Container> {
+  private String componentName;
+
+  public HomogeneityScorer(String componentName) {
+    this.componentName = componentName;
+  }
+
+  @Override
+  public boolean sortAscending() {
+    return false;
+  }
+
+  @Override
+  public double getScore(Container container) {
+    int totalComponentInstances = 0;
+    int totalInstances = container.getInstances().size();
+
+    if (totalInstances == 0) {
+      return 0;
+    }
+
+    for (PackingPlan.InstancePlan instancePlan : container.getInstances()) {
+      if (componentName.equals(instancePlan.getComponentName())) {
+        totalComponentInstances++;
+      }
+    }
+    return (double) totalComponentInstances / totalInstances;
```

Comment:
I think there is a confusion here. The imbalance metric you have will differentiate correctly between homogeneous and heterogeneous containers (e.g., [BBCD, BBB]). But it will not work well in a set of mixed containers. The reason is that usually the imbalance metric will try many times to remove from the most loaded container which mean that you reduce significantly the chances to remove containers overall. In your example with BBBBBBCD, ABC you have more chances getting rid of a container if you remove always from the least loaded one. Your solution currently does not remove multiple components at once. You remove one component at a time. In this setting, removing from the least loaded container will typically result in fewer containers overall. 

If you want to optimize scale down for imbalance then this approach is ok. But I personally believe that users would want to see resources been released when scaling down otherwise why would they even do it?